File: apps/backend/src/services/searchHistoryService.ts
import { supabase } from '../config/supabase';
import { SearchHistory } from '../types';

export class SearchHistoryService {
  async saveSearch(
    userId: string,
    query: string,
    recommendations: SearchHistory['recommendations']
  ): Promise<SearchHistory> {
    const { data, error } = await supabase
      .from('search_history')
      .insert({ user_id: userId, query, recommendations })
      .select()
      .single();

    if (error) {
      throw new Error(`Failed to save search: ${error.message}`);
    }

    return data;
  }

  async getSearchHistory(userId: string): Promise<SearchHistory[]> {
    const { data, error } = await supabase
      .from('search_history')
      .select('*')
      .eq('user_id', userId)
      .order('created_at', { ascending: false });

    if (error) {
      throw new Error(`Failed to fetch search history: ${error.message}`);
    }

    return data;
  }
}

File: apps/api/src/services/search_history_service.rs
use crate::error::AppError;
use crate::models::{Book, SearchHistory};
use anyhow::Result;
use serde_json::json;
use supabase_rs::SupabaseClient;

pub struct SearchHistoryService {
    client: SupabaseClient,
}

impl SearchHistoryService {
    pub fn new(client: SupabaseClient) -> Self {
        Self { client }
    }

    pub async fn save_search(
        &self,
        user_id: &str,
        query: &str,
        recommendations: &[Book],
    ) -> Result<SearchHistory> {
        let insert_data = json!({
            "user_id": user_id,
            "query": query,
            "recommendations": recommendations
        });

        let response = self
            .client
            .from("search_history")
            .insert(&insert_data)
            .select("*")
            .single()
            .execute()
            .await
            .map_err(|e| AppError::DatabaseError(e.to_string()))?;

        let search_history: SearchHistory = serde_json::from_value(response.data)
            .map_err(|e| AppError::SerializationError(e.to_string()))?;

        Ok(search_history)
    }

    pub async fn get_search_history(&self, user_id: &str) -> Result<Vec<SearchHistory>> {
        let response = self
            .client
            .from("search_history")
            .select("*")
            .eq("user_id", user_id)
            .order("created_at", Some("desc".to_string()))
            .execute()
            .await
            .map_err(|e| AppError::DatabaseError(e.to_string()))?;

        let history: Vec<SearchHistory> = serde_json::from_value(response.data)
            .map_err(|e| AppError::SerializationError(e.to_string()))?;

        Ok(history)
    }
}
